File: kotlin/src/main/kotlin/in/kanur/sorts/Merge.kt
package `in`.kanur.sorts

object Merge : Sort {
  private fun merge(left: IntArray, right: IntArray): IntArray {
    var leftIndex = 0
    var rightIndex = 0
    var merged = intArrayOf()

    while (leftIndex < left.size && rightIndex < right.size) {
      if (left[leftIndex] < right[rightIndex]) {
        merged += intArrayOf(left[leftIndex])
        leftIndex++
      } else {
        merged += intArrayOf(right[rightIndex])
        rightIndex++
      }
    }

    return merged +
      left.sliceArray(IntRange(leftIndex, left.size - 1)) +
      right.sliceArray(IntRange(rightIndex, right.size - 1))
  }

  override fun sort(list: IntArray): IntArray {
    if (list.size <= 1) {
      return list
    }

    val middle: Int = list.size / 2
    return merge(
      sort(list.sliceArray(IntRange(0, middle - 1))),
      sort(list.sliceArray(IntRange(middle, list.size - 1)))
    )
  }
}

File: typescript/src/merge.ts
function merge(left: number[], right: number[]) {
  let leftIndex = 0;
  let rightIndex = 0;
  let merged: number[] = [];

  while (leftIndex < left.length && rightIndex < right.length) {
    if (left[leftIndex] < right[rightIndex]) {
      merged = [...merged, left[leftIndex]];
      leftIndex++;
    } else {
      merged = [...merged, right[rightIndex]];
      rightIndex++;
    }
  }

  return [
    ...merged,
    ...left.slice(leftIndex, left.length),
    ...right.slice(rightIndex, right.length),
  ];
}

export function sort(numbers: number[]): number[] {
  if (numbers.length <= 1) {
    return numbers;
  }

  const middle = Math.floor(numbers.length / 2);
  return merge(
    sort(numbers.slice(0, middle)),
    sort(numbers.slice(middle, numbers.length))
  );
}
